docs/madr/decisions/046-meshservice-policy-matching.md

```diff
@@ -0,0 +1,174 @@
+# Policy matching with MeshService
+
+* Status: accepted
+
+Technical Story: #9707
+
+## Context and Problem Statement
+
+Now that MeshService exists, we need to enable users to apply policies in a way that works
+with the new `MeshService` resource. This includes using `StatefulSet`.
+Additionally, policy should be appliable to `MeshMultiZoneService`.
+The inspect API should also reflect policies that use `MeshService`
+which will be covered in a separate MADR.
+
+Some things to remember:
+
+* Currently `kind: MeshService` in `from[].targetRef` is deprecated.
+* While `kind: MeshService` in `spec.targetRef` points to both a Dataplane and a set of inbounds,
+  it is not a canonical reference.
+  Its effects effectively mix with any policies applied
+  to any `MeshService` that points to this `Dataplane`/inbound.
+  A canonical reference can only be a direct choice of an entire `Dataplane` object
+  and potentially a subset of inbounds.
+* A `MeshService` only has a 1-1 relationship when used as `to[].targetRef.kind`
+  since then it corresponds to a particular Envoy outbound listener/some number of clusters.
+
+## Decision Drivers
+
+* Targeting headless `MeshService`
+* Targeting `MeshMultiZoneService`
+* A balance between correctness and ease of use when configuring policy.
+
+## Considered Options
+
+This MADR adds 2 kinds `to[].targetRef.kind` and Mesh*Route `backendRefs[].kind`:
+
+* `MeshService`
+* `MeshMultiZoneService`
+
+as well as a number of new fields to the ref structure:
+
+* `port`
+* `namespace`
+* `labels`
+* `zone`
+
+This MADR doesn't address using `MeshService` as a reference anywhere else.
+
+Another option would be to have the `namespace` and `zone` fields under `labels` instead.
+The motivation for having them top level is that they have especially important
+semantic meaning, beyond other arbitrary, user-defined labels.
+
+## Decision Outcome
+
+These new kinds allow for selecting or directing traffic that's going to new,
+`MeshService`-based destinations.
+
+### `targetRef`
+
+The policy is applied
+on both the Envoy outbound listeners and clusters
+that correspond to the `MeshService`.
+
+References for `kind: MeshService` have the following structure:
+
+```yaml
+kind: MeshService
+name: backend
+labels: {}
+zone: east
+namespace: backend
+port: 80
+```
+
+In general, we try to do the expected thing and refer to `MeshService` via its display name.
```

Comment:
I'm tempted not to do this. While it may seem like better UX, I feel like this will introduce ambiguity, because you can use it to match a real name as well as display name.
I feel that labels even though more verbose it's more clear what is going on.

Especially that this way it's easier to then go and introduce those implicit things.

Alternatively I'd consider `displayName` field.

Reply:
Discussion:

```yaml
kind: Service
metadata:
  namespace: app-ns
  name: backend
---
kind: MeshService
metadata:
  namespace: app-ns
  name: backend
---
targetRef:
  kind: MeshService
  name: backend
  namespace: app-ns
```

```yaml
---
kind: MeshService
metadata:
  namespace: kube-system
  name: backend-a57694fd # hash namespace, zone and mesh
---
targetRef:
  kind: MeshService
  name: backend-a57694d
  namespace: app-ns
---
targetRef:
  kind: MeshService
  displayName: backend
  zone: other-zone
  namespace: app-ns
```